Preview: Enable published posts preview

aduth opened the pull request:
Closes (when #7130 is merged): #7179
Closes (when #7130 is merged): #2544
Related: #6257, #6882

**Note:** Merges to `fix/autosave-preview` (#7130) as base branch.
**Note:** As best I can tell, this branch is fully functional, but I have plans to include new end-to-end tests for various preview circumstances prior to merge.

This pull request seeks to allow the user to preview changes to a published post without first updating the post (i.e. the autosave revision preview).

In doing so, it...

- Refactors the `PostPreviewButton` component to avoid relying on a specific save workflow, and instead leverage the mere availability of a preview link
- Removes `preview_link` from the posts endpoints and adds instead to the autosaves endpoint (see #7179)

To the latter point, I landed at the conclusion that `preview_link` only makes sense in the context of autsosaves on consideration of:

- The post endpoints reflect the value of the canonical saved post. If there are no unsaved changes, the user should be directed to `post.link` as its preview.
- It does not belong in the revisions endpoints because it is not the case that any revision can be previewed. A preview only occurs for the latest revision of a post ([`_set_preview`](https://github.com/WordPress/WordPress/blob/6fd8080e7ee7599b36d4528f72a8ced612130b8c/wp-includes/revision.php#L556), [`wp_get_post_autosave`](https://github.com/WordPress/WordPress/blob/6fd8080e7ee7599b36d4528f72a8ced612130b8c/wp-includes/revision.php#L232-L246)).

Related Trac tickets:

- [REST API: Support autosaves](https://core.trac.wordpress.org/ticket/43316) (43316)
- [REST API: Expose 'preview_link' for viewable post types](https://core.trac.wordpress.org/ticket/44180) (44180)

Thus, if there are no unsaved changes for a post, the Preview button is effectively a link to `post.link`. If there are unsaved changes, the autosave is triggered and once its responses' `autosave.preview_link` becomes accessible, it is used in the redirect (regardless of whether or not the autosave occurred as a full save as described in #7124).

aduth committed:
REST API: Move preview_link to autosaves controller

aduth committed:
Post Preview: Redirect to preview by autosave property availability

aduth committed:
Preview Button: Use preview link as href if available

danielbachhuber commented:
> To the latter point, I landed at the conclusion that preview_link only makes sense in the context of autsosaves on consideration of:

This is an interesting idea. I hadn't considered it this way before, but it seems like a reasonable decision to me (particularly because I don't think core's existing implementation of previews is this well-considered).

danielbachhuber commented on editor/store/test/reducer.js:
```diff
@@ -2298,13 +2298,15 @@ describe( 'state', () => {
 						raw: 'The Excerpt',
 					},
 					status: 'draft',
+					preview_link: 'https://wordpress.org/?p=1',
```
Nit: Should this include `&preview=true` for clarity?

aduth replied:
>Nit: Should this include `&preview=true` for clarity?

Yep, sounds good.

danielbachhuber commented on lib/class-wp-rest-autosaves-controller.php:
```diff
@@ -339,6 +349,18 @@ public function prepare_item_for_response( $post, $request ) {
+
+		if ( ! empty( $schema['properties']['preview_link'] ) ) {
+			$response->data['preview_link'] = get_preview_post_link( $post->post_parent, array(
+				'preview_id'    => $post->post_parent,
```
Is `preview_id` supposed to be `$post->post_parent`? This ends up as `0` on a draft post:

<img width="997" alt="image" src="https://user-images.githubusercontent.com/36432/41064309-0c52699a-6990-11e8-8d32-5d5a5eeccd96.png">

aduth replied:
>Is `preview_id` supposed to be `$post->post_parent`? This ends up as `0` on a draft post:

Good call. We might consider using something like [`wp_is_post_autosave`](https://developer.wordpress.org/reference/functions/wp_is_post_autosave/) to vary the logic. Though in this context since we assume the `$post` is the result of either an autosave (`get_item`) or generated by an autosave intent, maybe just testing its type `$post->post_type === 'revision'`.

aduth replied:
Hmm the `-autosave` post name suffix checking in `wp_is_post_autosave` makes me think it might be better to rely on that. Guessing it might be possible for a non-autosave revision to be attached to a post parent?

danielbachhuber reviewed (commented):
Couple small nits. I also ran into an issue where the preview doesn't reflect the title:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/36432/41064451-67f30700-6990-11e8-923c-e30cec989732.png">

<img width="956" alt="image" src="https://user-images.githubusercontent.com/36432/41064459-6f38f1be-6990-11e8-9c4c-6b0f1014c79e.png">

It's not immediately obvious to me why that discrepancy exists.

aduth commented:
>Couple small nits. I also ran into an issue where the preview doesn't reflect the title:

I can reproduce this, and it seems to affect where we're previewing a post we'd already previewed once before in the same session:

1. Navigate to Posts > Add New
2. Enter a title
3. Hit preview
4. Return to editor tab
5. Change title
6. Hit preview

Expected: Preview is shown with updated title
Actual: Preview reflects original title.

The problem is that the second time we preview, we already have a `preview_link` available because we reset the autosave after the first preview. I think we either need to:

- Clear the `preview_link` before an autosave starts so it isn't considered for the popup window
- Consider change detection and invalidate the preview link as soon as the user makes an edit
- Force the autosave and interstitial screen if the post is dirty, even if a preview link is already available
   - Might be the simplest approach for now. Technically this is wasted for repeatedly hitting "Preview" on a published post, since it remains dirty after an autosave.

aduth commented:
Another complication: For repeated previews of a draft post, even if I make changes, the preview URL will never change (e.g. `?p=123&preview=true`), so it's hard to know that the preview is ready without either (re-)integrating awareness of the save process, or invalidating the preview link at some point (e.g. on edit or pre-autosave)

aduth committed:
Preview Button: Move window bind to after write

Was not reliably triggered when redirect occurs by setPreviewWindowLink, possibly because document.open triggered by document.write erased window handlers.

aduth committed:
Preview Button: Always trigger autosave when dirty

aduth committed:
Preview Button: Consider preview ready by change in value

aduth committed:
State: Set default value for savePost options

Allows use in handlers to assume options present

aduth committed:
Store: Clarify preview link with query parameter

aduth committed:
REST API: Set preview link by parent only if autosave

aduth committed:
Store: Invalid autosave preview link on save intent

Preview link is no longer valid until autosave completes, by which UI can infer availability of preview as being ready to view.

aduth committed:
Testing: Add preview E2E tests

aduth committed:
Testing: Fix copypasta

adamsilverstein commented:
Thanks for working on this @aduth - I'll give it a test and review as soon as possible!

aduth committed:
Preview: Use correct location property for URL comparison

aduth commented on editor/store/effects.js:
```diff
@@ -157,49 +157,14 @@ export default {
 				// draft or auto-draft.
 				const isRevision = newPost.id !== post.id;
 
-				// Thus, the following behaviors occur:
```
Note that these removed changes were introduced as part of the base branch to which this merges (#7130) so there is effectively no difference between this callback and what's in master at the moment.

aduth commented:
Going to merge this into the `fix/autosave-preview` branch (#7130) it has assigned as base, so as to be able to work on these related changes in one place.
